File: src/main/java/pw/switchcraft/plethora/gameplay/modules/RangeInfo.java
package pw.switchcraft.plethora.gameplay.modules;

import pw.switchcraft.plethora.api.reference.ConstantReference;
import pw.switchcraft.plethora.gameplay.modules.scanner.ScannerModuleItem;
import pw.switchcraft.plethora.gameplay.modules.sensor.SensorModuleItem;

import javax.annotation.Nonnull;
import java.util.function.IntUnaryOperator;

/**
 * Provides the range for a {@link SensorModuleItem} or {@link ScannerModuleItem}
 */
public interface RangeInfo extends ConstantReference<RangeInfo> {
    /**
     * The maximum range this module operates at.
     *
     * @return This module's range.
     */
    int getRange();

    /**
     * The cost for some bulk operation (sense/scan).
     *
     * @return The cost of a bulk operation.
     * @see pw.switchcraft.plethora.api.method.ICostHandler
     */
    int getBulkCost();

    @Nonnull
    @Override
    default RangeInfo get() {
        return this;
    }

    @Nonnull
    @Override
    default RangeInfo safeGet() {
        return this;
    }

    static RangeInfo of(int level, IntUnaryOperator cost, IntUnaryOperator range) {
        return new RangeInfo() {
            @Override
            public int getRange() {
                return range.applyAsInt(level);
            }

            @Override
            public int getBulkCost() {
                return cost.applyAsInt(level);
            }
        };
    }
}

File: src/main/java/io/sc3/plethora/gameplay/modules/RangeInfo.kt
package io.sc3.plethora.gameplay.modules

import io.sc3.plethora.api.reference.ConstantReference
import java.util.function.IntUnaryOperator

/**
 * Provides the range for a [SensorModuleItem] or [ScannerModuleItem]
 */
interface RangeInfo : ConstantReference<RangeInfo?> {
  /**
   * The maximum range this module operates at.
   *
   * @return This module's range.
   */
  val range: Int

  /**
   * The cost for some bulk operation (sense/scan).
   *
   * @return The cost of a bulk operation.
   * @see io.sc3.plethora.api.method.ICostHandler
   */
  val bulkCost: Int

  override fun get(): RangeInfo = this

  override fun safeGet(): RangeInfo = this

  companion object {
    fun of(level: Int, cost: IntUnaryOperator, range: IntUnaryOperator) = object : RangeInfo {
      override val range: Int
        get() = range.applyAsInt(level)
      override val bulkCost: Int
        get() = cost.applyAsInt(level)
    }
  }
}
